generator/lookml.py

```diff
@@ -129,42 +152,98 @@ def _lookml(namespaces, glean_apps, target_dir, dryrun, namespace_filter=[]):
 
+    run_generation = partial(_run_generation, metric_hub_repos)
+
+    generate_views = []
+    generate_datagroups = []
     v1_mapping = _glean_apps_to_v1_map(glean_apps)
+    logging.info("  Generating views and datagroups")
     for namespace, lookml_objects in _namespaces.items():
         if len(namespace_filter) == 0 or namespace in namespace_filter:
-            logging.info(f"\nGenerating namespace {namespace}")
-
             view_dir = target / namespace / "views"
             view_dir.mkdir(parents=True, exist_ok=True)
             views = list(
                 _get_views_from_dict(lookml_objects.get("views", {}), namespace)
             )
 
-            logging.info("  Generating views")
             v1_name: Optional[str] = v1_mapping.get(namespace)
-            for view_path in _generate_views(view_dir, views, v1_name, dryrun=dryrun):
-                logging.info(f"    ...Generating {view_path}")
+            for view in views:
+                generate_views.append(
+                    partial(
+                        _generate_view,
+                        view_dir,
+                        view,
+                        v1_name,
+                        dryrun,
+                    )
+                )
+                generate_datagroups.append(
+                    partial(
+                        generate_datagroup,
+                        view,
+                        target,
+                        namespace,
+                        dryrun,
+                    )
+                )
 
-            logging.info("  Generating datagroups")
-            generate_datagroups(views, target, namespace, dryrun=dryrun)
+    with Pool(parallelism) as pool:
+        pool.map(
+            run_generation,
+            generate_views + generate_datagroups,
+        )
 
+    generate_explores = []
+    for namespace, lookml_objects in _namespaces.items():
+        if len(namespace_filter) == 0 or namespace in namespace_filter:
+            view_dir = target / namespace / "views"
             explore_dir = target / namespace / "explores"
             explore_dir.mkdir(parents=True, exist_ok=True)
             explores = lookml_objects.get("explores", {})
-            logging.info("  Generating explores")
-            for explore_path in _generate_explores(
-                explore_dir, namespace, explores, view_dir, v1_name
-            ):
-                logging.info(f"    ...Generating {explore_path}")
+            v1_name = v1_mapping.get(namespace)
+            generate_explores += [
+                partial(
+                    _generate_explore,
+                    explore_dir,
+                    namespace,
+                    explore_name,
+                    explore,
+                    view_dir,
+                    v1_name,
+                )
+                for explore_name, explore in explores.items()
+            ]
+
+    logging.info("  Generating explores")
+    with Pool(parallelism) as pool:
```

Comment:
I'd suggest trying to use one pool for everything for better efficiency (no need to repeatedly tear down old processes and spin up new sets of processes for explores and dashboards), and building up the lists of things to generate in one pass would also avoid the repeated namespace loop logic.

For max efficiency you could throw everything at the pool at once, or to keep things a bit more orderly each set of artifacts could be processed in turn, like
```python
    with Pool(parallelism) as pool:
        logging.info("  Generating views")
        pool.map(run_generation, generate_views)
        logging.info("  Generating datagroups")
        pool.map(run_generation, generate_datagroups)
        ...
```

Reply:
Good point. I changed the logic. It still is faster to combine running the view and data group generation as they don't depend on each other (unlike explore and dashboard generation which read view lookml).